File: src/0020-Valid-Parentheses/0020.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <stack>
using namespace std;
static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    bool isValid(string s) 
    {
        stack<char> mystack;
        unordered_map<char, char> match = {{'(',')'}, {'{','}'}, {'[',']'}};
        for (auto& i : s)
        {
            if (i == '(' || i == '{' || i == '[')
            {
                mystack.push(i);
            }
            else
            {
                if (mystack.empty()) return false;
                char top = mystack.top();
                mystack.pop();
                if (match[top] != i) return false;
            }
        }
        
        return mystack.size() == 0;
    }
};
int main()
{
    string str = "()";
    cout << Solution().isValid(str) << endl;
    return 0;
}

File: src/0020-Valid-Parentheses/0020.java
package leetcodeChecks;

import java.util.*;
public class oo2o {
    class Solution {
        public boolean isValid(String s) {

            if(s.isEmpty()){
                return true;
            }

            Stack<Character> st = new Stack<Character>();
            for(int i=0; i < s.length();i++){

                if(s.charAt(i)=='(' || s.charAt(i)=='[' ||s.charAt(i)=='{'){
                    st.push(s.charAt(i));
                }
                else{
                    if(st.isEmpty()){
                        return false;
                    }
                    char c= st.peek();


                    if(s.charAt(i)==')' && c=='('|| s.charAt(i) == ']' && c=='['|| s.charAt(i) =='}' && c=='{'){
                        st.pop();
                    }else{
                        return false;
                    }
                }
            }

            if(st.isEmpty()){
                return true;
            }
            else{
                return false;
            }


        }
    }
}
